File: 26_01_2025_Maximum_Employees_to_Be_Invited_to_a_Meeting.py


class Solution:
    def maximumInvitations(self, favorite):
        n = len(favorite)
        in_degree = [0] * n
        longest_chain = [0] * n
        visited = [False] * n

        for f in favorite:
            in_degree[f] += 1

        queue = deque()
        for i in range(n):
            if in_degree[i] == 0:
                queue.append(i)

        while queue:
            front = queue.popleft()
            visited[front] = True
            nxt = favorite[front]
            longest_chain[nxt] = longest_chain[front] + 1

            in_degree[nxt] -= 1
            if in_degree[nxt] == 0:
                queue.append(nxt)

        max_cycle = 0
        total = 0

        for i in range(n):
            if not visited[i]:
                c = i
                length = 0
                while not visited[c]:
                    visited[c] = True
                    c = favorite[c]
                    length += 1

                if length == 2:
                    total += (2 + longest_chain[i] + longest_chain[favorite[i]])
                else:
                    max_cycle = max(max_cycle, length)

        return max(total, max_cycle)

File: 26_01_2025_Maximum_Employees_to_Be_Invited_to_a_Meeting.cpp
class Solution {
public:
    int maximumInvitations(vector<int>& favorite) {

        int n = favorite.size();
        vector<int> inDegree(n, 0);
        vector<int> longestChain(n, 0);
        vector<bool> visited(n, false);
        
        // Step 1: Calculate in-degrees for all nodes
        for (int i : favorite) {
            inDegree[i]++;
        }

        // Step 2: Topological Sort (Kahn's Algorithm)
        queue<int> q;
        for (int i = 0; i < n; i++) {
            if (inDegree[i] == 0) {
                q.push(i);
            }
        }

        // Step 3: Process each node in topological order to update the longest chain
        while (!q.empty()) {
            int front = q.front();
            q.pop();
            visited[front] = true;
            int next = favorite[front];
            longestChain[next] = longestChain[front] + 1;

            if (--inDegree[next] == 0) {
                q.push(next);
            }
        }

        // Step 4: Detect cycles and compute cycle sizes
        int maxCycle = 0, total = 0;

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                int c = i, len = 0;
                while (!visited[c]) {
                    visited[c] = true;
                    c = favorite[c];
                    len++;
                }

                if (len == 2) {
                    total += (2 + longestChain[i] + longestChain[favorite[i]]);
                } else {
                    maxCycle = max(maxCycle, len);
                }
            }
        }

        // Step 5: Return the maximum of the largest cycle size and two-node cycle sum
        return max(total, maxCycle);
    }
};

